fastcrypto-vdf/src/class_group/discriminant.rs

```diff
@@ -1,29 +1,51 @@
 use crate::math::parameterized_group::Parameter;
 use fastcrypto::error::FastCryptoError::InvalidInput;
 use fastcrypto::error::{FastCryptoError, FastCryptoResult};
+use lazy_static::lazy_static;
 use num_bigint::{BigInt, ToBigInt};
 use num_integer::Integer;
-use num_traits::Signed;
+use num_traits::{One, Signed};
 use serde::{Deserialize, Serialize};
 use std::ops::Neg;
+use std::str::FromStr;
 
 /// A discriminant for an imaginary class group. The discriminant is a negative integer congruent to
 /// 1 mod 8.
-#[derive(PartialEq, Eq, Debug, Clone, Serialize, Deserialize)]
+#[derive(PartialEq, Eq, Debug, Clone, Serialize)]
 pub struct Discriminant(#[serde(with = "crate::class_group::bigint_serde")] BigInt);
 
+lazy_static! {
+    /// Fixed 3072 bit discriminant. Generated from the seed [1,2,3] using [Discriminant::from_seed].
+    // TODO: Generate this using a seed that we provably cannot influence.
+    pub static ref DISCRIMINANT_3072: Discriminant = Discriminant(BigInt::from_str("-4080390101490206102067801750685552291425412528983716161454985565795560716833845004659207152503580931176637478422335625954692628868126419714053340412299850300602673802493259771830686596468801304317015718872352674945215883546019961626928140286675493693757393881479657605888983279619347902770789061953207866325747708864327315769009839190765716943013935708854055658243676903245686125751909996824976354309908771869043784640567352757672203749399825983258156684652782580603170228640173640869773628592618889352385821753919281706169861276929330689892675986265846043432389737049521845230769417696140636288030698887830215613149485135897148487896368642774768920061430225392365148291796645740474628778185683682893521776342856643134668770656709308404166182149870849376649591338267281149794078240401323227967073641261327798339424740171219484355109588337730742391198073121589465833677609362668436116144203312494461735357918360857667357985711").unwrap());
+}
+
+impl<'de> Deserialize<'de> for Discriminant {
+    fn deserialize<D>(deserializer: D) -> Result<Discriminant, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        Discriminant::try_from(bigint_serde::deserialize(deserializer)?)
+            .map_err(serde::de::Error::custom)
+    }
+}
+
 impl TryFrom<BigInt> for Discriminant {
     type Error = FastCryptoError;
 
-    /// A valid discriminant should be a negative prime congruent to 1 mod 8. The sign and
-    /// congruency are checked here but the primality is _not_. See also [Discriminant::check_primality].
+    /// A valid discriminant should be a negative prime congruent to 1 mod 8. For large discriminants,
+    /// this is very slow to check.
     fn try_from(value: BigInt) -> FastCryptoResult<Self> {
-        if !value.is_negative() || value.mod_floor(&BigInt::from(8)) != BigInt::from(1) {
+        if !value.is_negative()
+            || value.mod_floor(&BigInt::from(8)) != BigInt::one()
+            || !is_probable_prime(value.abs().magnitude())
```

Comment:
Another thing to be careful when sampling is "non determinism". We generally avoid probabilistic functions that could have non deterministic behavior. I guess this function won't be exposed to the blockchain api. If it does, we need to ensure someone cannot cause a fork.

Reply:
I've actually disabled the random trials for the Miller-Rabin test (see the `is_probable_prime` functions) to ensure this, but I'll go over it again to be ceratin.